core/src/main/scala/org/broadinstitute/dsde/rawls/entities/compact/batch/BatchHandling.scala

```diff
@@ -119,23 +212,26 @@ trait BatchHandling extends LazyLogging with AttributeSupport {
         if (notFoundReferenceTargets.nonEmpty) {
           throw generateReferenceError("Could not resolve some entity references", notFoundReferenceTargets)
         }
-        // did we find all the reference sources? This should never happen, but let's be defensive
-        val notFoundReferenceSources = allReferences.keys.toSet diff actuallyFound
-        if (notFoundReferenceSources.nonEmpty)
-          throw generateReferenceError("Could not resolve some entity sources", notFoundReferenceSources)
+        // did we find all the entities we originally wrote? This should never happen, but let's be defensive
+        val notFoundWrittenEntities = writtenEntityRefs diff actuallyFound
+        if (notFoundWrittenEntities.nonEmpty)
+          throw generateReferenceError("Could not resolve some entity sources", notFoundWrittenEntities)
       }
 
       // build a lookup table for the ids we found
       idLookup: Map[AttributeEntityReference, Long] = foundIds.map { rec =>
         rec.toAttributeEntityReference -> rec.id
       }.toMap
 
-      // rehydrate the looked-up ids into sources and targets (from_id, to_id)
+      // rehydrate the looked-up ids for the entities we wrote in this batch
+      referenceSourcesToDelete: Set[Long] = writtenEntityRefs.map(ref => idLookup(ref))
+      // rehydrate the looked-up ids into sources and targets for the references to insert (from_id, to_id)
       referencesToInsert: Set[RefPointers] = allReferences.map { case (from, tos) =>
         val toIds = tos.map(idLookup).toSet
         RefPointers(idLookup(from), toIds)
       }.toSet
       // insert the references into the ENTITY_REFS table.
+      _ <- repository.queries.deleteAllReferencesFrom(referenceSourcesToDelete)
       _ <- repository.queries.upsertReferences(referencesToInsert)
```

Comment:
I don't love this blunt delete-all then insert-all approach, but doing anything fancier was much more complex.